### PR TITLE
Update RemoveSnapshots to protect cherry-picked data.

### PR DESCRIPTION
**Pull request by mehtaashish23**
Rebased version of #802 

This updates RemoveSnapshots to detect cherry-picked commits and skip cleanup.
This doesn't handle complex cases where snapshots are explicitly expired using ExpireSnapshots.expireSnapshotId, or where the current snapshot is directly modified. We will want a more thorough follow-up for those cases, tracked by #744.

This does handle cases where a change is picked into the current table state and the original commit expires. Cleanup for picked changes is done when the commit in the current snapshot's ancestors expires.

### CODE REVIEW

File: core/src/test/java/org/apache/iceberg/TestTables.java [hunk 1]
```diff
@@ -218,9 +222,7 @@ public OutputFile newOutputFile(String path) {
 
     @Override
     public void deleteFile(String path) {
-      if (!new File(path).delete()) {
-        throw new RuntimeIOException("Failed to delete file: " + path);
-      }
+      DELETED_FILE_PATHS.add(path);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestTables.java]: Instead of updating a static variable that we need to clean up between test cases, how about using the delete callback? I think that's going to be a better way to test.

File: api/src/main/java/org/apache/iceberg/ExpireSnapshots.java [hunk 2]
```diff
@@ -40,11 +40,13 @@
...
 
   /**
-   * Expires a specific {@link Snapshot} identified by id.
+   * Expires a specific {@link Snapshot} identified by id. This API isn't compatible with
+   * cherry-pick functionality (issue #744), and hence marked Deprecated.
    *
    * @param snapshotId long id of the snapshot to expire
    * @return this for method chaining
    */
+  @Deprecated
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/ExpireSnapshots.java]: Can you move this to a separate PR? I don't think we need to update RemoveSnapshots and deprecate a method at the same time.

File: core/src/main/java/org/apache/iceberg/RemoveSnapshots.java [hunk 3]
```diff
@@ -208,12 +226,40 @@ private void cleanExpiredFiles(List<Snapshot> snapshots, Set<Long> validIds, Set
...
+          continue;
+        }
+
+        long sourceSnapshotId = PropertyUtil.propertyAsLong(
+            snapshot.summary(), SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP, -1);
+        if (ancestorIds.contains(sourceSnapshotId)) {
+          // this commit was cherry-picked from a commit that is in the current table state. do not clean up its
+          // changes because it would revert data file additions that are in the current table.
+          // A -- B
+          //  `- C (source=B) <-- this commit
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/RemoveSnapshots.java]: Quick note: this case needs a follow-up because we could expire A and B, then later expire C (not in the current table state) and not be able to detect that we should not delete its files. Expiring B means that we need to also expire C.

File: core/src/test/java/org/apache/iceberg/TestWapWorkflow.java [hunk 4]
```diff
@@ -684,4 +685,87 @@ public void testNonWapCherrypick() {
           table.manageSnapshots().cherrypick(firstSnapshotId).commit();
         });
   }
+
+  @Test
+  public void testWithExpiringWapThenCherrypick() {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestWapWorkflow.java]: I think these tests should be in the suite for snapshot expiration, not the suite for WAP. Can you move them?

File: core/src/test/java/org/apache/iceberg/TestWapWorkflow.java [hunk 5]
```diff
@@ -684,4 +685,87 @@ public void testNonWapCherrypick() {
...
+    table.newAppend()
+            .appendFile(FILE_A)
+            .commit();
+    // first WAP commit
+    table.newAppend()
+            .appendFile(FILE_B)
+            .set(SnapshotSummary.STAGED_WAP_ID_PROP, "123456789")
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestWapWorkflow.java]: The tests shouldn't use WAP properties at all. Correctness depends on where commits were picked and should ignore the WAP ID. The reason is that two commits with the same WAP id don't contain the same files. They are logically the same data, not physically the same files. The only equivalence that matters is when a snapshot itself has been picked because the files are physically the same.

File: api/src/main/java/org/apache/iceberg/ExpireSnapshots.java (hunk 2 above)
Comment (rominparekh) [reply api/src/main/java/org/apache/iceberg/ExpireSnapshots.java]: Also can we possibly provide a message to use an alternate API for deprecation. Something like - 
```java
@Deprecate("Use `expireOlderThan()` method instead")
```

File: core/src/test/java/org/apache/iceberg/TestTables.java (hunk 1 above)
Comment (rominparekh) [reply core/src/test/java/org/apache/iceberg/TestTables.java]: +1

File: core/src/test/java/org/apache/iceberg/TestWapWorkflow.java (hunk 5 above)
Comment (rominparekh) [reply core/src/test/java/org/apache/iceberg/TestWapWorkflow.java]: @mehtaashish23 - Can you add a UT for each scenario we are trying to handle? It might help debug in case we see rogue behavior. Thanks.

File: core/src/main/java/org/apache/iceberg/RemoveSnapshots.java (hunk 3 above)
Comment (mehtaashish23) [reply core/src/main/java/org/apache/iceberg/RemoveSnapshots.java]: NOTE: I think we can handle this case, i.e. when picking a snapshot for expiration, we should also pick its source snapshot, but considering the gotchas, I would recommend to keep the API simple with expireOlderThan. Also our use case is to use `expireOlderThan` API for now.

I will log a ticket for this.

**Comment by mehtaashish23** [comment]
@rdblue , @rominparekh Added the delete callback than static variable dependency in test code. Please review. 
Also Ryan pointed out a scenario, which doesn't protect cherrypick work from expireSnapshotById. Let me know, if I should fix in this PR, or as pointed, I can log the ticket, and pick that up explicitly.

File: core/src/test/java/org/apache/iceberg/TableTestBase.java [hunk 6]
```diff
@@ -109,8 +113,8 @@ public void cleanupTables() {
         !name.startsWith("snap") && Files.getFileExtension(name).equalsIgnoreCase("avro")));
   }
 
-  private TestTables.TestTable create(Schema schema, PartitionSpec spec) {
-    return TestTables.create(tableDir, "test", schema, spec);
+  private TestTables.TestTable create(Schema schema, PartitionSpec spec, Boolean accumulateDeletedFiles) {
+    return TestTables.create(tableDir, "test", schema, spec, accumulateDeletedFiles);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TableTestBase.java]: This class doesn't need to change at all because you can use [`ExpireSnapshots.deleteWith`](https://iceberg.apache.org/javadoc/0.7.0-incubating/org/apache/iceberg/ExpireSnapshots.html#deleteWith-java.util.function.Consumer-) and pass a callback. In each test case, you'd do this to get the list of deletes:

```java
  List<String> deletedFiles = Lists.newArrayList();
  table.expireSnapshots()
      .deleteWith(deletedFiles::add)
      .expireOlderThan(t)
      .commit();
```

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 7]
```diff
@@ -387,4 +393,55 @@ public void dataFilesCleanup() throws IOException {
...
+  public void testWithExpiringWapThenCherrypick() {
+    // first WAP commit
+    table.newAppend()
+            .appendFile(FILE_B)
+            .stageOnly()
+            .commit();
+
+    TableMetadata base = readMetadata();
+    // pick the snapshot that's staged but not committed
+    Snapshot wap1Snapshot = base.snapshots().get(0);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: This shouldn't use the name `wap`, since this is independent of the WAP pattern.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 8]
```diff
@@ -387,4 +393,55 @@ public void dataFilesCleanup() throws IOException {
     Assert.assertTrue("FILE_A should be deleted", deletedFiles.contains(FILE_A.path().toString()));
     Assert.assertTrue("FILE_B should be deleted", deletedFiles.contains(FILE_B.path().toString()));
   }
+  @Test
+  public void testWithExpiringWapThenCherrypick() {
+    // first WAP commit
+    table.newAppend()
+            .appendFile(FILE_B)
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: Continuing indentation should be 4 spaces from the start of the previous line, not 8.

**Comment by rdblue** [comment]
Thanks for working on this, @mehtaashish23! Could you remove references to WAP from the tests? Also, I agree that there should be a test case for each cherry-pick case handled by the delete code.

File: core/src/test/java/org/apache/iceberg/TableTestBase.java (hunk 6 above)
Comment (mehtaashish23) [reply core/src/test/java/org/apache/iceberg/TableTestBase.java]: ohh really. Thanks, nice API!

File: core/src/main/java/org/apache/iceberg/RemoveSnapshots.java [hunk 9]
```diff
@@ -208,12 +226,40 @@ private void cleanExpiredFiles(List<Snapshot> snapshots, Set<Long> validIds, Set
...
+          continue;
+        }
+
+        if (pickedAncestorSnapshotIds.contains(sourceSnapshotId)) {
+          // this commit was cherry-picked from a commit that is in the current table state. do not clean up its
+          // changes because it would revert data file additions that are in the current table.
+          // A -- C -- E (source=B)
+          //  `- B `- D (source=B) <-- this commit
```
Comment (mehtaashish23) [inline core/src/main/java/org/apache/iceberg/RemoveSnapshots.java]: Having D (cherry pick of B) isn't possible, if B is already an ancestor, encountered this while writing UTs.

**Comment by mehtaashish23** [comment]
@rdblue Thanks for the inputs, incorporated your feedback, and adding 3rd UT to capture all the cases, and clearly mentioned on each UT, what scenario it is trying to execute, in similar way you added comment in code. LMK

**Comment by mehtaashish23** [comment]
@rdblue : Do you any concern? I was thinking if we could merge, and I can rebase/cherry-pick this change onto our internal fork.

File: core/src/main/java/org/apache/iceberg/RemoveSnapshots.java (hunk 9 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/RemoveSnapshots.java]: The case that would cause this is:
* A is committed
* B is staged
* C is committed
* B is picked to create D
* The table is rolled back to C
* B is picked to create E

The main thing is the rollback, which can cause alternate branches with the same commit picked. I would expect in practice to have other commits (like one between C and D that is removed by the rollback) but we don't need those to test this case.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 10]
```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
...
+    snapshotB.addedFiles().forEach(item -> {
+      Assert.assertTrue(deletedFiles.contains(item.path().toString()));
+    });
+  }
+
+  /**
+   * Expire cherry-pick the commit as shown below, when `B` is in table's current state
+   *  Table:
+   *  A - B
+   *   `- C (B) <--this commit
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: The label "this commit" makes sense in context for `RemoveSnapshots` because it is iterating through snapshots/commits. It makes less sense here. Instead, I would use "current" and note which commits are expiring (e.g., "expiring A and B").

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 11]
```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
...
+   */
+  @Test
+  public void testWithExpiringDanglingStageCommit() {
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` staged commit
+    table.newAppend()
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: Can you change this to an overwrite and also delete FILE_A? We should also test that deleted files are not physically deleted.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 12]
```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
...
+    List<String> deletedFiles = new ArrayList<>();
+
+    // Expire all commits including dangling staged snapshot.
+    table.expireSnapshots()
+        .deleteWith(deletedFiles::add)
+        .expireOlderThan(snapshotB.timestampMillis() + 1)
+        .commit();
+
+    // Files should be deleted of dangling staged snapshot
+    snapshotB.addedFiles().forEach(item -> {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: I think it would be more clear to use a `Set<String>` and validate using `assertEquals`. That way it verifies no additional files were deleted.

```java
Set<String> expectedDeletes = snapshotB.addedFiles().stream()
    .map(added -> added.path().toString())
    .collect(Collectors.toSet());
Assert.assertEquals("Should delete only files added in snapshot B", expectedDeletes, deletedFiles);
```

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java (hunk 10 above)
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: I think it would also be easier to understand if this explicitly labeled the source with `(source=B)` instead of just `(B)`.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 13]
```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
...
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .commit();
+    // pick the snapshot 'B`
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: Nit: should have a newline before this comment.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 14]
```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
...
+    // Move the table back to `B`
+    table.manageSnapshots()
+        .setCurrentSnapshot(snapshotB.snapshotId())
+        .commit();
+    List<String> deletedFiles = new ArrayList<>();
+
+    // Expire `C`
+    table.expireSnapshots()
+        .deleteWith(deletedFiles::add)
+        .expireSnapshotId(snapshotC.snapshotId())
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: You can simplify this and avoid using `expireSnapshotId` by using `expireOlderThan` a future time (`Long.MAX_VALUE`). Snapshot B won't expire because it is the current snapshot.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 15]
```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
...
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    // pick the snapshot that's staged but not committed
+    TableMetadata base = readMetadata();
+    Snapshot snapshotB = base.snapshots().get(1);
+
+    // `C` commit to let cherry-pick take effect
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: It wasn't clear at first why this was needed, but it is to avoid using a fast-forward commit instead of a cherry-pick. Can you update the comment in `RemoveSnapshots` since the case there isn't possible?

**Comment by rdblue** [comment]
Thanks @mehtaashish23! I think we're almost ready to commit this.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java (hunk 12 above)
Comment (mehtaashish23) [reply core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: +1, it would be good to add exact files, after considering following.
1. Apart from dataFile, we do delete manifests as well for dangling snapshots
2. The order in which the files are deleted, shouldn't matter, so I have handling that as well.

I think, we should have couple of test to check manifest deletes in case of normal remove snapshots API. Let me know, I can log a ticket explicitly for that, if you agree.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java (hunk 15 above)
Comment (mehtaashish23) [reply core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: Yeah, I have also changed another UT, which was not taking care of fast forward.

**Comment by mehtaashish23** [comment]
@rdblue accommodated your feedback, with making sure, we check the exact file expected in clean up. Please review. It took time, as I was working on ongoing commitment.

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java [hunk 16]
```diff
@@ -387,4 +389,190 @@ public void dataFilesCleanup() throws IOException {
...
+    expectedDeletes.add(snapshotB.manifestListLocation());
+    snapshotB.manifests().forEach(file -> {
+      //Only the manifest of B should be deleted.
+      if (file.snapshotId() == snapshotB.snapshotId()) {
+        expectedDeletes.add(file.path());
+      }
+    });
+    Assert.assertSame("Files deleted count should be expected", expectedDeletes.size(), deletedFiles.size());
+    //Take the diff
+    expectedDeletes.removeAll(deletedFiles);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: Why does this remove files from the set instead of using `assertEqual`?

**Comment by rdblue** [comment]
@mehtaashish23, I think the tests are doing the right thing and I want to get this into master so that cherry-picked commits are no longer dangerous. We can solve the remaining issues in a follow-up. Thanks!

File: core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java (hunk 16 above)
Comment (mehtaashish23) [reply core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: The order was coming to be different, and I didn't wanted to rely on order.

Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java]: Comparing two sets would not rely on order.
